File: ion/src/com/koushikdutta/ion/InputStreamParser.java
package com.koushikdutta.ion;

import com.koushikdutta.async.DataEmitter;
import com.koushikdutta.async.DataSink;
import com.koushikdutta.async.callback.CompletedCallback;
import com.koushikdutta.async.future.Future;
import com.koushikdutta.async.parser.AsyncParser;
import com.koushikdutta.async.parser.ByteBufferListParser;
import com.koushikdutta.async.stream.ByteBufferListInputStream;

import java.io.InputStream;
import java.lang.reflect.Type;

/**
 * Created by koush on 11/3/13.
 */
class InputStreamParser implements AsyncParser<InputStream> {
    @Override
    public Future<InputStream> parse(DataEmitter emitter) {
        return new ByteBufferListParser().parse(emitter)
                .thenConvert(from -> new ByteBufferListInputStream(from));
    }

    @Override
    public void write(DataSink sink, InputStream value, CompletedCallback completed) {
        throw new AssertionError("not implemented");
    }

    @Override
    public Type getType() {
        return InputStream.class;
    }

    @Override
    public String getMime() {
        return null;
    }
}

File: ion/src/com/koushikdutta/ion/util/InputStreamParser.kt
package com.koushikdutta.ion.util

import com.koushikdutta.scratch.AsyncInput
import com.koushikdutta.scratch.AsyncRead
import com.koushikdutta.scratch.Promise
import com.koushikdutta.scratch.buffers.ByteBufferListInputStream
import com.koushikdutta.scratch.http.AsyncHttpMessageContent
import com.koushikdutta.scratch.stream.createAsyncInput
import java.io.InputStream

class InputStreamParser(override val contentType: String = "application/octet-stream") : AsyncParser<InputStream> {
    override fun parse(read: AsyncRead): Promise<InputStream> = ByteBufferListParser().parse(read).then {
        ByteBufferListInputStream(it)
    }

    override val type = InputStream::class.java
}

class InputStreamSerializer(val contentLength: Long? = null, val contentType: String = "application/octet-stream") : AsyncSerializer<InputStream> {
    override fun write(value: InputStream) = Promise<AsyncHttpMessageContent> {
        val input = value.createAsyncInput()
        object : AsyncHttpMessageContent, AsyncInput by input {
            override val contentLength = this@InputStreamSerializer.contentLength
            override val contentType = this@InputStreamSerializer.contentType
        }
    }
}
